### PR TITLE
ARROW-2467: [Rust] Add generated IPC code

### PR DESCRIPTION
**Pull request by andygrove**
I generated the IPC code after installing the flatbuffers release available via conda but still had to manually modify the code to make it compile due to these issues with flatbuffers:

https://github.com/google/flatbuffers/issues/5052
https://github.com/google/flatbuffers/issues/5053
https://github.com/google/flatbuffers/issues/5054

I think this is the best we can do for now.

### CODE REVIEW

File: rust/arrow/src/ipc/gen/mod.rs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#![allow(non_snake_case)]
+
+pub mod File;
```
Comment (nevi-me) [inline rust/arrow/src/ipc/gen/mod.rs]: Do the 3 issues stop it from compiling altogether? When I worked on feather support eariler in the year, I didn't have any compilation issues (unless it's issues in the sense of CI being strict), but I had to still work around the nested modules.

Do you mind sharing the `flatc` script/command that you used when building? We should probably document that somewhere.

I think in the near future, we might need to add a `build.rs` to auto-generate fbs and proto files, I can do that as I want to add support for a Flight client in the next release.

Comment (andygrove) [reply rust/arrow/src/ipc/gen/mod.rs]: I just pushed a commit to add instructions to the README so you can see the steps I took.

Comment (nevi-me) [reply rust/arrow/src/ipc/gen/mod.rs]: Thanks, I'll take a look later today

**Review by nevi-me — APPROVED** [review]
Happy with this as a start, once flatbuffer issues are resolved upstream, we can consider adding a build script for flatc, though it might not be necessary as the `*.fbs` files are rarely changed.

An ideal future is where we have:

```rust
// [arrow/rust/arrow/build.rs]

use std::path::Path;
use flatc_rust;

fn main() {
    flatc_rust::run(flatc_rust::Args {
        lang: "rust",
        inputs: &[
            Path::new("../../format/FIle.fbs"),
            Path::new("../../format/Message.fbs"),
            Path::new("../../format/Schema.fbs"),
            Path::new("../../format/Tensor.fbs"),
            Path::new("../../format/SparseTensor.fbs"),
        ],
        out_dir: Path::new("./src/ipc/gen/"),
        // doesn't seem to be honoured
        includes: &[Path::new("../../format/")],
        ..Default::default()
    }).expect("Unable to build flatbuffer files");
}

// [arrow/rust/arrow/Cargo.toml]
[package]
...
build = "build.rs"

...

[build-dependencies]
flatc-rust = "0.1"

```

File: rust/arrow/src/ipc/gen/mod.rs (hunk 1 above)
Comment (nevi-me) [reply rust/arrow/src/ipc/gen/mod.rs]: I ended up taking longer, my apologies

**Comment by andygrove** [comment]
@sunchao @paddyhoran @kszucs Please take a look when you have time

**Review by kszucs — APPROVED** [review]
LGTM, created an issue for generating the code automatically once the upstream issues are resolved: https://issues.apache.org/jira/browse/ARROW-5135

**Comment by kszucs** [comment]
I'm merging it, thanks @andygrove!

Do We have any missing pieces to add rust to the format integration tests https://issues.apache.org/jira/browse/ARROW-3690?

Rust implementation grows pretty quickly and We should ensure that it is compatible with the rest of the implementations.

**Comment by paddyhoran** [comment]
> Rust implementation grows pretty quickly and We should ensure that it is compatible with the rest of the implementations.

Agreed, I plan to start looking at this when I get a chance.

**Comment by kszucs** [comment]
Cool, thanks @paddyhoran. The first step is to implement the appropiate commands on rust side, then I can help with adding it to the python testing script.

**Pull request closed without merging**